pyrealm/phenology/fapar_limitation.py

```diff
@@ -0,0 +1,415 @@
+        self.annual_mean_ca = annual_mean_ca
+        """Ambient CO2 partial pressure. [Pa]"""
+        self.annual_mean_chi = annual_mean_chi
+        """Annual mean ratio of leaf-internal CO2 partial pressure to c_a during the 
+        growing season (>0℃). [Pa]"""
+        self.annual_mean_vpd = annual_mean_vpd
+        """Annual mean vapour pressure deficit (VPD) during the growing season [Pa]"""
+        self.annual_total_precip = annual_total_precip
+        """Annual total precipitation. [mol m^{-2} year^{-1}]"""
+        self.aridity_index = aridity_index
+        """Climatological estimate of local aridity index."""
+
+        self._check_shapes()
+
+        # Constants used for phenology computations
+        self.phenology_const = PhenologyConst()
+
+        #  f_0 is the ratio of annual total transpiration of annual total
+        #  precipitation, which is an empirical function of the climatic Aridity Index
+        #  (AI).
+        a, b, c = self.phenology_const.f0_coefficients
+        f_0 = a * np.exp(-b * np.log(aridity_index / c) ** 2)
+
+        fapar_energylim = 1.0 - self.phenology_const.z / (
+            self.phenology_const.k * annual_total_potential_gpp
+        )
+        fapar_waterlim = (
+            f_0
+            * annual_total_precip
+            * annual_mean_ca
+            * (1 - annual_mean_chi)
+            / (1.6 * annual_mean_vpd * annual_total_potential_gpp)
+        )
+
+        self.fapar_max = np.minimum(fapar_waterlim, fapar_energylim)
+        """Maximum fapar given water or energy limitation."""
+        self.energy_limited = fapar_energylim < fapar_waterlim
+        """Is fapar_max limited by water or energy."""
+        self.annual_precip_molar = annual_total_precip
+        """The annual precipitation in moles."""
+
+        self.lai_max = -(1 / self.phenology_const.k) * np.log(1.0 - self.fapar_max)
+
+    @classmethod
+    def from_pmodel(
+        cls,
+        pmodel: PModel,
+        growing_season: NDArray[np.bool],
+        datetimes: NDArray[np.datetime64],
+        precip: NDArray[np.float64],
+        aridity_index: NDArray[np.float64],
+        gpp_penalty_factor: NDArray[np.float64],
+    ) -> Self:
+        r"""Get FaparLimitation from PModel input.
+
+        Computes the input for fAPAR_max from the P Model and additional inputs.
+
+        Args:
+            pmodel: pyrealm.pmodel.PModel
+            growing_season: Bool array indicating which times are within growing
+              season by some definition and implementation.
+            datetimes: Array of datetimes to consider.
+            precip: Precipitation for given datetimes.
+            aridity_index: Climatological estimate of local aridity index.
+            gpp_penalty_factor: Penalty factor to be applied to pmodel.gpp
+        """
+
+        check_datetimes(datetimes)
+
+        annual_total_potential_gpp = get_annual(
+            pmodel.gpp * gpp_penalty_factor, datetimes, growing_season, "total"
+        )
```

Comment:
A few things here:
* GPP (and precipitation below) shouldn't use the growing season filtering - they use the actual totals across all observations.
* GPP in the PModel is reported in µg C m-2 s-1, but this model requires an annual estimate of mol C m-2 year-1.  I think the easiest approach here is to calculate the mean of the GPP values within the year and then scale it up by the number of seconds in the year and then divide by the molar mass of Carbon (which  will be in `PModel.core_const.k_c_molmass`.

Reply:
Just to double-check, this should be totals, not mean values, for GPP, right?